capi/include/hyper.h

```diff
@@ -0,0 +1,545 @@
+
+ This takes ownership of the `hyper_body *`, you must not use it or
+ free it after setting it on the request.
+ */
+hyper_code hyper_request_set_body(hyper_request *req, hyper_body *body);
+
+/*
+ Free an HTTP response after using it.
+ */
+void hyper_response_free(hyper_response *resp);
+
+/*
+ Get the HTTP-Status code of this response.
+
+ It will always be within the range of 100-599.
+ */
+uint16_t hyper_response_status(const hyper_response *resp);
+
+/*
+ Get a pointer to the reason-phrase of this response.
+
+ This buffer is not null-terminated.
+
+ This buffer is owned by the response, and should not be used after
+ the response has been freed.
+
+ Use `hyper_response_reason_phrase_len()` to get the length of this
+ buffer.
+ */
+const uint8_t *hyper_response_reason_phrase(const hyper_response *resp);
+
+/*
+ Get the length of the reason-phrase of this response.
+
+ Use `hyper_response_reason_phrase()` to get the buffer pointer.
+ */
+size_t hyper_response_reason_phrase_len(const hyper_response *resp);
+
+/*
+ Get the HTTP version used by this response.
+
+ The returned value could be:
+
+ - `HYPER_HTTP_VERSION_1_0`
+ - `HYPER_HTTP_VERSION_1_1`
+ - `HYPER_HTTP_VERSION_2`
+ - `HYPER_HTTP_VERSION_NONE` if newer (or older).
+ */
+int hyper_response_version(const hyper_response *resp);
+
+/*
+ Gets a reference to the HTTP headers of this response.
+
+ This is not an owned reference, so it should not be accessed after the
+ `hyper_response` has been freed.
+ */
+hyper_headers *hyper_response_headers(hyper_response *resp);
+
+/*
+ Take ownership of the body of this response.
+
+ It is safe to free the response even after taking ownership of its body.
+ */
+hyper_body *hyper_response_body(hyper_response *resp);
+
+/*
+ Iterates the headers passing each name and value pair to the callback.
+
+ The `userdata` pointer is also passed to the callback.
+
+ The callback should return `HYPER_ITER_CONTINUE` to keep iterating, or
+ `HYPER_ITER_BREAK` to stop.
+ */
+void hyper_headers_foreach(const hyper_headers *headers,
+                           hyper_headers_foreach_callback func,
+                           void *userdata);
+
+/*
+ Sets the header with the provided name to the provided value.
+
+ This overwrites any previous value set for the header.
+ */
+hyper_code hyper_headers_set(hyper_headers *headers,
+                             const uint8_t *name,
+                             size_t name_len,
+                             const uint8_t *value,
+                             size_t value_len);
+
+/*
+ Adds the provided value to the list of the provided name.
+
+ If there were already existing values for the name, this will append the
+ new value to the internal list.
+ */
+hyper_code hyper_headers_add(hyper_headers *headers,
+                             const uint8_t *name,
+                             size_t name_len,
+                             const uint8_t *value,
+                             size_t value_len);
```

Comment:
Question: when I add a request header *without* a value, with `value_len` set to zero, Hyper seems to still insert a space after the colon. This means I can't produce an outgoing header without a trailing space if there's no value for the header. Is this intended or something that could be considered to be changed? It's a little detail that makes it difficult for me to generate the HTTP request with Hyper as flexible as I can with curl (and it's a wasted byte over the wire).

Reply:
Huh. That must be because the encoder is essentially a loop of `buf += name + ": " + val + "\r\n"`. I suppose we could add a check if the value is empty. Doesn't seem like something that is very common, but I understand curl already supports it.

Reply:
There are web services "out there" that wants a header without value, and yes curl has specific support to provide such a header. In our test suite we verify the entire protocol "dump" as issued by the client - which in my case now subsequently makes the curl+hyper client issue an extra space for this case and therefore the test case fails...

Reply:
Yea, figured. Added a check.